gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/CommitActivityImpl.java

```diff
@@ -0,0 +1,183 @@
+      // TODO: Task state dir is a stub with the assumption it is always colocated with the workunits dir (as in the case of MR which generates workunits)
+      Path jobIdParent = new Path(workSpec.getWorkUnitsDir()).getParent();
+      Path jobOutputPath = new Path(new Path(jobIdParent, "output"), jobIdParent.getName());
+      log.info("Output path at: " + jobOutputPath + " with fs at " + fs.getUri());
+      StateStore<TaskState> taskStateStore = Help.openTaskStateStore(workSpec, fs);
+      Collection<TaskState> taskStateQueue =
+          ImmutableList.copyOf(
+              TaskStateCollectorService.deserializeTaskStatesFromFolder(taskStateStore, jobOutputPath, numDeserializationThreads));
+      commitTaskStates(jobState, taskStateQueue, globalGobblinContext);
+
+    } catch (Exception e) {
+      //TODO: IMPROVE GRANULARITY OF RETRIES
+      throw ApplicationFailure.newNonRetryableFailureWithCause(
+          "Failed to commit dataset state for some dataset(s) of job <jobStub>",
+          IOException.class.toString(),
+          new IOException(e),
+          null
+      );
+    }
+
+    return 0;
+  }
+
+  protected FileSystem loadFileSystemForUri(URI fsUri, State stateConfig) throws IOException {
+    return HadoopUtils.getFileSystem(fsUri, stateConfig);
+  }
+
+  void commitTaskStates(State jobState, Collection<TaskState> taskStates, JobContext jobContext) throws IOException {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = createDatasetStatesByUrns(taskStates);
+    final boolean shouldCommitDataInJob = shouldCommitDataInJob(jobState);
+    final DeliverySemantics deliverySemantics = DeliverySemantics.AT_LEAST_ONCE;
+    final int numCommitThreads = 1;
+
+    if (!shouldCommitDataInJob) {
+      log.info("Job will not commit data since data are committed by tasks.");
+    }
+
+    try {
+      if (!datasetStatesByUrns.isEmpty()) {
+        log.info("Persisting dataset urns.");
+      }
+
+      List<Either<Void, ExecutionException>> result = new IteratorExecutor<>(Iterables
+          .transform(datasetStatesByUrns.entrySet(),
+              new Function<Map.Entry<String, JobState.DatasetState>, Callable<Void>>() {
+                @Nullable
+                @Override
+                public Callable<Void> apply(final Map.Entry<String, JobState.DatasetState> entry) {
+                  return new SafeDatasetCommit(shouldCommitDataInJob, false, deliverySemantics, entry.getKey(),
+                      entry.getValue(), false, jobContext);
+                }
+              }).iterator(), numCommitThreads,
+          ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of("Commit-thread-%d")))
+          .executeAndGetResults();
+
+      IteratorExecutor.logFailures(result, null, 10);
+
+      if (!IteratorExecutor.verifyAllSuccessful(result)) {
+        // TODO: propagate cause of failure
+        throw new IOException("Failed to commit dataset state for some dataset(s) of job <jobStub>");
```

Comment:
suggest to include more specific info (e.g. of the task state store path)

...basically something to correlate it w/ output prior from `logFailures`

Reply:
It's logged in the previous area `IteratorExecutor.logFailures` which cuts off after a set number of error logs to prevent overcrowding logs